slick/src/main/scala/za/co/absa/fadb/slick/SlickFunction.scala

```diff
@@ -16,64 +16,96 @@
-  * @tparam I - The input type of the function
-  * @tparam R - The return type of the function
-  */
-trait SlickFunction[I, R] extends DBFunctionFabric {
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
 
+/**
+ *  Base class for Slick DB functions.
+ *
+ *  @tparam I the input type of the function
+ *  @tparam R the result type of the function
+ */
+private[slick] trait SlickFunctionBase[I, R] {
   /**
-    * A reference to the [[SlickPgEngine]] to use the [[za.co.absa.fadb.DBFunction DBFunction]] with
-    */
-  implicit val dbEngine: SlickPgEngine
+   *  The `GetResult[R]` instance used to read the query result into `R`.
+   */
+  protected def slickConverter: GetResult[R]
 
   /**
-    * This is expected to return SQL part of the [[SlickQuery]] (eventually returned by the `SlickPgFunction.query` function
-    * @param values - the values to pass over to the database function
-    * @return       - the Slick representation of the SQL
-    */
+   *  Generates a Slick `SQLActionBuilder` representing the SQL query for the function.
+   *
+   *  @param values the input values for the function
+   *  @return the Slick `SQLActionBuilder` representing the SQL query
+   */
   protected def sql(values: I): SQLActionBuilder
 
+  def fieldsToSelect: Seq[String]
+}
+
+private[slick] trait SlickFunction[I, R] extends SlickFunctionBase[I, R] {
   /**
-    * This is expected to return a method to convert the [[slick.jdbc.PositionedResult slick.PositionedResult]], the Slick general SQL result
-    * format into the `R` type
-    * @return - the converting function
-    */
-  protected def slickConverter: GetResult[R]
+   *  Generates a `SlickQuery[R]` representing the SQL query for the function.
+   *
+   *  @param values the input values for the function
+   *  @return the `SlickQuery[R]` representing the SQL query
+   */
+  protected def query(values: I): SlickQuery[R] = new SlickQuery(sql(values), slickConverter)
+}
+
+private[slick] trait SlickFunctionWithStatus[I, R] extends SlickFunctionBase[I, R] {
+  /**
+   * Generates a `SlickQueryWithStatus[R]` representing the SQL query for the function with status support.
+   *
+   * @param status - the status to check
+   * @return - Success or failure the status means
+   */
+  protected def query(values: I): SlickQueryWithStatus[R] =
+    new SlickQueryWithStatus[R](sql(values), slickConverter, checkStatus)
+
+  // Expected to be mixed in by an implementation of StatusHandling
+  def checkStatus[A](statusWithData: FunctionStatusWithData[A]): Either[StatusException, A]
+}
+
+object SlickFunction {
+  /**
+   *  Class for Slick DB functions with status support.
+   */
+  abstract class SlickSingleResultFunctionWithStatus[I, R](functionNameOverride: Option[String] = None)(implicit
+    override val schema: DBSchema,
+    DBEngine: SlickPgEngine
```

Comment:
```suggestion
    dBEngine: SlickPgEngine
```

also elsewhere, just to keep the convention

Reply:
fixed